paymentapi/src/main/java/fi/hel/verkkokauppa/payment/logic/builder/PaytrailPaymentContextBuilder.java

```diff
@@ -0,0 +1,105 @@
+import org.springframework.web.reactive.function.client.WebClient;
+
+import fi.hel.verkkokauppa.payment.service.ServiceConfigurationClient;
+
+@Component
+public class PaytrailPaymentContextBuilder {
+
+    private static final Logger LOG = LoggerFactory.getLogger(PaytrailPaymentContextBuilder.class);
+
+    @Autowired
+    private Environment env;
+
+    @Autowired
+    private ServiceConfigurationClient serviceConfigurationClient;
+
+    @Autowired
+    private CommonServiceConfigurationClient commonServiceConfigurationClient;
+
+
+    public PaytrailPaymentContext buildFor(String namespace, String merchantId) {
+        PaytrailPaymentContext defaultContext = new PaytrailPaymentContext();
+        defaultContext.setNamespace(namespace);
+
+        // set default values
+        defaultContext.setDefaultCurrency(env.getRequiredProperty("payment_default_paytrail_currency"));
+        defaultContext.setDefaultLanguage(env.getRequiredProperty("payment_default_paytrail_language"));
+        defaultContext.setReturnUrl(env.getRequiredProperty("payment_default_paytrail_return_url"));
+        defaultContext.setNotifyUrl(env.getRequiredProperty("payment_default_paytrail_notify_url"));
+        defaultContext.setAggregateMerchantId(env.getRequiredProperty("payment_default_paytrail_aggregate_merchant_id"));
+        defaultContext.setCp(env.getRequiredProperty("payment_default_paytrail_cp"));
+
+        // fetch namespace and merchant specific service configuration from mapping api
+        PaytrailPaymentContext enrichedContext = null;
+        try {
+            enrichedContext = enrichWithNamespaceConfiguration(defaultContext);
+            enrichedContext = enrichWithMerchantConfiguration(enrichedContext, merchantId);
+        } catch (Exception e) {
+            LOG.error("failed to fetch service configuration for namespace {} and merchant {}: " + namespace, merchantId, e);
+            // TODO by default allowing payments with defaults
+        }
+
+        if (enrichedContext != null) {
+            LOG.debug("using merchant specific service configuration");
+            return enrichedContext;
+        } else {
+            LOG.debug("using default service configuration");
+            return defaultContext;
+        }
+    }
+
+    private JSONObject getNamespaceConfiguration(String namespace) {
+        WebClient client = serviceConfigurationClient.getClient();
+        JSONObject namespaceServiceConfiguration = serviceConfigurationClient.getAllServiceConfiguration(client, namespace);
+
+        // TODO caching
+
+        return namespaceServiceConfiguration;
+    }
+
+    private PaytrailPaymentContext enrichWithNamespaceConfiguration(PaytrailPaymentContext context) {
+        JSONObject namespaceServiceConfiguration = getNamespaceConfiguration(context.getNamespace());
+
+        String returnUrl = (String) namespaceServiceConfiguration.get("payment_return_url");
+        if (returnUrl != null)
```

Comment:
Could you add braces to these if statements, I know you just copy pasted these but still :D

Reply:
Sure thing, I'll add them.